### PR TITLE
Tests for long task observer: https://github.com/wicg/longtasks

### PR DESCRIPTION
**Pull request by spanicker**
@igrigorik PTAL

### CODE REVIEW

**Comment by spanicker** [comment]
Comments on test structuring for parent/child iframes (and cross origin) is appreciated. Thanks.

**Comment by spanicker** [comment]
@tdresser for comment

File: longtask-timing/resources/subframe-with-longtask.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE HTML>
+<meta charset=utf-8>
+<meta name="viewport" content="width=device-width">
+
+<title>Long Task Iframe</title>
+<h1>Long Task in Inline Script</h1>
+
+<script>
+    var begin = window.performance.now();
+    while (window.performance.now() < begin + 100);
```
Comment (tdresser) [inline longtask-timing/resources/subframe-with-longtask.html]: We use 51ms in places - why use 100 here?

File: longtask-timing/resources/makelongtask.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+/* Generate a slow task */
```
Comment (tdresser) [inline longtask-timing/resources/makelongtask.js]: Missing .

File: longtask-timing/resources/subframe-observing-longtask.html [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+<meta charset=utf-8>
+<meta name="viewport" content="width=device-width">
+<title>Child Iframe</title>
+<h1>Child Iframe observing long tasks</h1>
+
+<script>
+    var observer = new PerformanceObserver(function(entryList) {
+      for (i = 0; i < entryList.getEntries().length; i++) {
+      	var longtask = entryList.getEntries()[i];
+      	 // ignore long task generated within here, as part of making this iframe.
```
Comment (tdresser) [inline longtask-timing/resources/subframe-observing-longtask.html]: Capitalize

**Review by tdresser — APPROVED** [review]
This all looks pretty reasonable to me (I left a couple nits). I don't have thoughts on how to handle the cross process iframe case. @RByers might have thoughts?

File: longtask-timing/longtask-attributes.html [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        t.step_func(function (entryList) {
+            var entries = entryList.getEntries();
+            assert_equals(entries.length, 1,
+                "Exactly one entry is expected.");
+            var longtask = entries[0];
+            assert_equals(longtask.entryType, "longtask");
+            assert_equals(longtask.name, "same-origin-self");
+            assert_greater_than(longtask.duration, 50);
+            assert_equals(longtask.startTime, Math.floor(longtask.startTime),
+                "startTime expected to have 1 miillisecond granularity");
```
Comment (igrigorik) [inline longtask-timing/longtask-attributes.html]: millisecond typo

File: longtask-timing/longtask-attributes.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+            // Assert the TaskAttributionTiming entry in attribution.
+            assert_equals(longtask.attribution.length, 1,
+                "Exactly one attribution entry is expected");
+            var attribution = longtask.attribution[0];
+            assert_equals(attribution.entryType, "taskattribution");
```
Comment (igrigorik) [inline longtask-timing/longtask-attributes.html]: Hrm, should we be hyphenating these? We do in `name`.

File: longtask-timing/longtask-in-childiframe-crossorigin.html [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        t.step_func(function (entryList) {
+            var entries = entryList.getEntries();
+            assert_equals(entries.length, 1,
+                "Exactly one entry is expected.");
+            var longtask = entries[0];
+            assert_equals(longtask.entryType, "longtask");
+            assert_equals(longtask.name, "cross-origin-descendant");
+            assert_greater_than(longtask.duration, 50);
+            assert_equals(longtask.startTime, Math.floor(longtask.startTime),
+                "startTime expected to have 1 miillisecond granularity");
```
Comment (igrigorik) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: same typo

File: longtask-timing/longtask-in-childiframe-crossorigin.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            observer.disconnect();
+            t.done();
+        })
+    );
+    observer.observe({entryTypes: ["longtask"]});
+
+}, "Performance longtask entries in child iframe are observable in parent");
+</script>
+
+<!--TODO(panicker): simulate cross-origin instead -->
```
Comment (igrigorik) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: Hmm? Confused by the TODO.. What's left for this one?

**Review by igrigorik — COMMENTED** [review]
- s/miillisecond/millisecond
- meta: should 'taskattribution' be 'task-attribution' for readability and consistency?

File: longtask-timing/longtask-in-childiframe-crossorigin.html [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            assert_equals(longtask.attribution.length, 1,
+                "Exactly one attribution entry is expected");
+            var attribution = longtask.attribution[0];
+            assert_equals(attribution.entryType, "taskattribution");
+            assert_equals(attribution.name, "frame");
+            assert_equals(attribution.duration, 0);
+            assert_equals(attribution.startTime, 0);
+            assert_equals(attribution.frameId, "child-iframe-id");
+            assert_equals(attribution.frameName, "child-iframe-name");
+            assert_equals(attribution.frameSrc, "https://wicg.github.io/longtasks/randomslow.html");
```
Comment (RByers) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: I think web-platform-tests are run in an environment that prevents arbitrary network access, but they [do have a number of different origins available](https://github.com/w3c/web-platform-tests#running-the-tests).  See, for example, [this test](https://github.com/w3c/web-platform-tests/blob/master/html/semantics/embedded-content/the-iframe-element/cross_origin_parentage.html).

**Review by RByers — CHANGES_REQUESTED** [review]

File: longtask-timing/longtask-attributes.html (hunk 4 above)
Comment (spanicker) [inline longtask-timing/longtask-attributes.html]: Rounding sounds better, but looking at existing code for "clamping resolution" I'm finding floor.
Filed crbug.com/681718 to track this.

Comment (spanicker) [inline longtask-timing/longtask-attributes.html]: Done.

File: longtask-timing/longtask-attributes.html (hunk 5 above)
Comment (spanicker) [inline longtask-timing/longtask-attributes.html]: This is consistent with `entryType` - we don't hyphenate "longtask" for instance. Should we?

File: longtask-timing/longtask-in-childiframe-crossorigin.html (hunk 6 above)
Comment (spanicker) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: Fixed

File: longtask-timing/longtask-in-childiframe-crossorigin.html (hunk 8 above)
Comment (spanicker) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: This is exactly what I was looking for. Thanks!

File: longtask-timing/longtask-in-childiframe-crossorigin.html (hunk 7 above)
Comment (spanicker) [inline longtask-timing/longtask-in-childiframe-crossorigin.html]: Updated with Rick's pointer, we shouldn't try to load actual external sites in tests

File: longtask-timing/resources/makelongtask.js (hunk 2 above)
Comment (spanicker) [inline longtask-timing/resources/makelongtask.js]: Fixed

File: longtask-timing/resources/subframe-observing-longtask.html (hunk 3 above)
Comment (spanicker) [inline longtask-timing/resources/subframe-observing-longtask.html]: Fixed

File: longtask-timing/resources/subframe-with-longtask.html (hunk 1 above)
Comment (spanicker) [inline longtask-timing/resources/subframe-with-longtask.html]: Good catch, fixed.

**Comment by spanicker** [comment]
Friendly ping.

**Review by igrigorik — APPROVED** [review]

**Comment by igrigorik** [comment]
Looks good! @RByers any other feedback from your end, before we merge?

**Comment by tobie** [comment]
Ping @wpt-pr-bot.

**Comment by tobie** [comment]
Sorry for the noise, folks.

**Comment by spanicker** [comment]
Friendly ping!

**Comment by igrigorik** [comment]
@RByers ptal :-)

**Comment by RByers** [comment]
Sorry for the delay - I'm now on a flight and so can't look but please
don't block on me landing these tests!

**Comment by igrigorik** [comment]
👍

**Comment by annevk** [comment]
This PR broke master somehow. @jgraham @jugglinmike @bobholt

**Comment by spanicker** [comment]
Apologies. Too much time elapsed between this CL and merge.
Which tests specifically are failing?
Should I fix or rollback ?

**Comment by annevk** [comment]
It got fixed by @jgraham in https://github.com/w3c/web-platform-tests/commit/4a28c0eb39b8e217c74bf66da769286745b35e38. What you might want to do though is see if you can use test.step_timeout and remove yourself from the lint exclusion resource though.

**Comment by spanicker** [comment]
Ok I'll follow up on that. Thanks!

**Pull request merged**